### PR TITLE
[ESM] support attention API

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Addresses https://github.com/huggingface/transformers/pull/34954 and updates ESM to supports attention API and modeling outputs

### CODE REVIEW

**Comment by zucchini-nlp** [comment]
run-slow: esm

**Comment by zucchini-nlp** [comment]
run-slow: esm

**Comment by zucchini-nlp** [comment]
run-slow: esm, evolla

File: src/transformers/models/evolla/modeling_evolla.py [hunk 1]
```diff
@@ -274,39 +300,35 @@ def forward(
         head_mask: Optional[torch.FloatTensor] = None,
         encoder_hidden_states: Optional[torch.FloatTensor] = None,
         encoder_attention_mask: Optional[torch.FloatTensor] = None,
-        output_attentions: Optional[bool] = False,
+        **kwargs: Unpack[TransformersKwargs],
     ) -> tuple[torch.Tensor]:
-        hidden_shape = (hidden_states.shape[0], -1, self.num_attention_heads, self.attention_head_size)
+        batch_size, seq_length = hidden_states.shape[:-1]
+        hidden_shape = (batch_size, -1, self.num_attention_heads, self.attention_head_size)
```
Comment (vasqu) [inline src/transformers/models/evolla/modeling_evolla.py]: Please use `(batch_size, seq_len, -1, self.attention_head_size)` - the -1 is important for TP as the number of heads are sliced by the tp degree

File: src/transformers/models/evolla/modeling_evolla.py [hunk 2]
```diff
@@ -274,39 +300,35 @@ def forward(
...
 
+        relative_position_scores = None
         if self.position_embedding_type == "rotary":
             query_layer, key_layer = self.rotary_embeddings(query_layer, key_layer)
+        elif self.position_embedding_type == "relative_key" or self.position_embedding_type == "relative_key_query":
+            if self.config._attn_implementation != "eager":
+                raise ValueError(
+                    f"ESM {self.config._attn_implementation} attention does not support {self.position_embedding_type} embeddings. "
+                    "Set attention explicitly to 'eager' with `model.set_attn_implementation('eager')`"
+                )
```
Comment (vasqu) [inline src/transformers/models/evolla/modeling_evolla.py]: We already thought about deprecating this, https://github.com/huggingface/transformers/pull/38301#discussion_r2176948603

We have an easier time here as it has no cross dep to bert

File: src/transformers/models/evolla/modeling_evolla.py [hunk 3]
```diff
@@ -315,38 +337,30 @@ def forward(
 
             if self.position_embedding_type == "relative_key":
                 relative_position_scores = torch.einsum("bhld,lrd->bhlr", query_layer, positional_embedding)
-                attention_scores = attention_scores + relative_position_scores
             elif self.position_embedding_type == "relative_key_query":
                 relative_position_scores_query = torch.einsum("bhld,lrd->bhlr", query_layer, positional_embedding)
                 relative_position_scores_key = torch.einsum("bhrd,lrd->bhlr", key_layer, positional_embedding)
-                attention_scores = attention_scores + relative_position_scores_query + relative_position_scores_key
+                relative_position_scores = relative_position_scores_query + relative_position_scores_key
```
Comment (vasqu) [inline src/transformers/models/evolla/modeling_evolla.py]: Let's move this to eager instead like #38301 (if we decide to still support this)

File: src/transformers/models/evolla/modeling_evolla.py [hunk 4]
```diff
@@ -315,38 +337,30 @@ def forward(
...
 
-        outputs = (context_layer, attention_probs) if output_attentions else (context_layer,)
+        attn_output, attn_weights = attention_interface(
+            self,
+            query_layer,
+            key_layer,
+            value_layer,
+            attention_mask,
+            dropout=0.0 if not self.training else self.dropout,
+            scaling=self.scaling,
```
Comment (vasqu) [inline src/transformers/models/evolla/modeling_evolla.py:359]: `self.scaling` is not defined in init, no?

File: src/transformers/models/evolla/modeling_evolla.py [hunk 5]
```diff
@@ -514,21 +409,19 @@ def forward(
         head_mask=None,
         encoder_hidden_states=None,
         encoder_attention_mask=None,
-        output_attentions=False,
-        cache_position=None,
+        **kwargs: Unpack[TransformersKwargs],
     ):
         hidden_states_ln = self.LayerNorm(hidden_states)
-        self_outputs = self.self(
+        attn_output, _ = self.self(
```
Comment (vasqu) [inline src/transformers/models/evolla/modeling_evolla.py:417]: Shouldn't we update the Pretrained module to allow recording outputs? So the `_can_record_outputs`

Comment (vasqu) [reply src/transformers/models/evolla/modeling_evolla.py:417]: Also updating the `_supports_xxx` flags

File: src/transformers/models/esm/modeling_esm.py [hunk 6]
```diff
@@ -734,6 +597,16 @@ class EsmPreTrainedModel(PreTrainedModel):
     _no_split_modules = ["EsmLayer", "EsmFoldTriangularSelfAttentionBlock", "EsmEmbeddings"]
     _keys_to_ignore_on_load_unexpected = ["position_embeddings.weight"]
     _supports_flash_attn = True
+    _supports_sdpa = True
+    _supports_attention_backend = True
```
Comment (vasqu) [inline src/transformers/models/esm/modeling_esm.py:603]: As long as we have encoder-decoder capabilities I'd be careful with this as we would need different kwargs for encoder/decoder/cross attention - had this planned but havent gotten to it yet

File: src/transformers/models/esm/modeling_esm.py [hunk 7]
```diff
@@ -841,43 +712,31 @@ def forward(
...
         # we need to make broadcastable to [batch_size, num_heads, seq_length, seq_length]
         if self.config.is_decoder and encoder_hidden_states is not None:
             encoder_batch_size, encoder_sequence_length, _ = encoder_hidden_states.size()
             encoder_hidden_shape = (encoder_batch_size, encoder_sequence_length)
             if encoder_attention_mask is None:
-                encoder_attention_mask = torch.ones(encoder_hidden_shape, device=device)
+                encoder_attention_mask = torch.ones(encoder_hidden_shape, device=inputs_embeds.device)
             encoder_extended_attention_mask = self.invert_attention_mask(encoder_attention_mask)
         else:
             encoder_extended_attention_mask = None
```
Comment (vasqu) [inline src/transformers/models/esm/modeling_esm.py:744]: Can we align mask creation to the same as in #38301

It will make refactoring easier and the mask creations are "more proven".

**Review by vasqu — COMMENTED** [review]
Some initial thoughts, it would be nice if we could align things with #38301 as we can then refactor things more easily afterwards.

File: src/transformers/models/evolla/modeling_evolla.py (hunk 2 above)
Comment (zucchini-nlp) [reply src/transformers/models/evolla/modeling_evolla.py]: ig it's going to be deprecated later, doesn't seem related to this PR's goal

File: src/transformers/models/evolla/modeling_evolla.py (hunk 4 above)
Comment (zucchini-nlp) [reply src/transformers/models/evolla/modeling_evolla.py:359]: could be, if modular didn't copy it correctly. Will check

File: src/transformers/models/esm/modeling_esm.py (hunk 6 above)
Comment (zucchini-nlp) [reply src/transformers/models/esm/modeling_esm.py:603]: Agreed, at this point tho the model has only encoder defined in model code which supports attention API. I think it's safe enough

File: src/transformers/models/esm/modeling_esm.py (hunk 7 above)
Comment (zucchini-nlp) [reply src/transformers/models/esm/modeling_esm.py:744]: yeah, also thought of it, but after seeing it supports non-causal mask as well I think we need a cleaner approach for that in the future. Some kind of a small function that would decide which mask to construct and return it

File: src/transformers/models/evolla/modeling_evolla.py (hunk 5 above)
Comment (zucchini-nlp) [reply src/transformers/models/evolla/modeling_evolla.py:417]: yeah, I assumed modular copied everything, apparently not

**Comment by zucchini-nlp** [comment]
run-slow: esm, evolla

File: tests/test_modeling_common.py [hunk 8]
```diff
@@ -4391,7 +4391,6 @@ def test_flash_attention_2_continue_generate_with_position_ids(self):
                 next_token_logits_from_generate = generation_out.logits[-1]
 
                 # acceptable numerical instability
-                # print(next_token_logits_from_generate, next_token_logits)
```
Comment (vasqu) [inline tests/test_modeling_common.py:4394]: :laughing:

File: src/transformers/models/evolla/configuration_evolla.py [hunk 9]
```diff
@@ -76,7 +76,6 @@ def __init__(
         initializer_range=0.02,
         layer_norm_eps=1e-05,
         position_embedding_type="rotary",
-        use_cache=True,
```
Comment (vasqu) [inline src/transformers/models/evolla/configuration_evolla.py:79]: Hmm, why was this removed?

File: src/transformers/utils/generic.py [hunk 10]
```diff
@@ -974,22 +974,22 @@ def check_model_inputs(func):
 
     @wraps(func)
     def wrapper(self, *args, **kwargs):
-        use_cache = kwargs.get("use_cache")
-        if use_cache is None:
-            use_cache = getattr(self.config, "use_cache", False)
+        use_cache = (
```
Comment (vasqu) [inline src/transformers/utils/generic.py:977]: I assume this change comes from the removal of `us_cache` in evolla?

File: src/transformers/models/evolla/modeling_evolla.py [hunk 11]
```diff
@@ -229,11 +221,67 @@ def forward(self, q: torch.Tensor, k: torch.Tensor) -> tuple[torch.Tensor, torch
...
+    attn_weights = torch.matmul(query, key.transpose(2, 3)) * scaling
+
+    if attention_mask is not None:
+        causal_mask = attention_mask[:, :, :, : key.shape[-2]]
+        attn_weights = attn_weights + causal_mask
+
+    attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(query.dtype)
+    attn_weights = nn.functional.dropout(attn_weights, p=dropout, training=module.training)
+
+    if hasattr(module, "position_embedding_type") and module.position_embedding_type in [
```
Comment (vasqu) [inline src/transformers/models/evolla/modeling_evolla.py:243]: The attribute should exist in any case no?

Comment (vasqu) [reply src/transformers/models/evolla/modeling_evolla.py:243]: I think you can just copy https://github.com/huggingface/transformers/blob/e0f1e83d3ff876e0671a2f50e2543af2a9823433/src/transformers/models/bert/modeling_bert.py#L213-L231

File: src/transformers/models/evolla/modeling_evolla.py [hunk 12]
```diff
@@ -229,11 +221,67 @@ def forward(self, q: torch.Tensor, k: torch.Tensor) -> tuple[torch.Tensor, torch
...
+        positional_embedding = positional_embedding.to(dtype=query.dtype)  # fp16 compatibility
+
+        if module.position_embedding_type == "relative_key":
+            relative_position_scores = torch.einsum("bhld,lrd->bhlr", query, positional_embedding)
+        elif module.position_embedding_type == "relative_key_query":
+            relative_position_scores_query = torch.einsum("bhld,lrd->bhlr", query, positional_embedding)
+            relative_position_scores_key = torch.einsum("bhrd,lrd->bhlr", key, positional_embedding)
+            relative_position_scores = relative_position_scores_query + relative_position_scores_key
+
+        attn_weights = attn_weights + relative_position_scores
```
Comment (vasqu) [inline src/transformers/models/evolla/modeling_evolla.py:261]: The order is wrong, it should be before the scaling and after the matmul (but with scaling == 1 it's not important :eyes:)

File: src/transformers/models/evolla/modeling_evolla.py [hunk 13]
```diff
@@ -229,11 +221,67 @@ def forward(self, q: torch.Tensor, k: torch.Tensor) -> tuple[torch.Tensor, torch
...
+
+    if hasattr(module, "position_embedding_type") and module.position_embedding_type in [
+        "relative_key",
+        "relative_key_query",
+    ]:
+        if module.config._attn_implementation != "eager":
+            raise ValueError(
+                f"ESM {module.config._attn_implementation} attention does not support {module.position_embedding_type} embeddings. "
+                "Set attention explicitly to 'eager' with `model.set_attn_implementation('eager')`"
+            )
```
Comment (vasqu) [inline src/transformers/models/evolla/modeling_evolla.py]: This seems like a dead path, the error should occur before like https://github.com/huggingface/transformers/blob/e0f1e83d3ff876e0671a2f50e2543af2a9823433/src/transformers/models/bert/modeling_bert.py#L318-L325

**Review by vasqu — COMMENTED** [review]
Oops, commented on the modular generated file but should be carried over to esm - my bad ^^'

File: src/transformers/models/esm/modeling_esm.py (hunk 7 above)
Comment (vasqu) [reply src/transformers/models/esm/modeling_esm.py:744]: Not meaning the attention mask interface yet as that definitely needs an update for non-causal variants :D thinking of https://github.com/huggingface/transformers/blob/e0f1e83d3ff876e0671a2f50e2543af2a9823433/src/transformers/models/bert/modeling_bert.py#L990-L1034 
(without the checks on the dims). It would make it easy to remove all those functions later on if we keep it consistent across models that have yet to get the mask interface. (And I have tested it quite thoroughly on all attention variations)

File: src/transformers/models/evolla/configuration_evolla.py (hunk 9 above)
Comment (zucchini-nlp) [reply src/transformers/models/evolla/configuration_evolla.py:79]: The model's forward doesn't use cache. Otherwise `record_outputs` passes `use_cache` in forward which errors out

File: src/transformers/utils/generic.py (hunk 10 above)
Comment (zucchini-nlp) [reply src/transformers/utils/generic.py:977]: Not all models use cache, e.g. evolla yes. If we want to use record outputs with non-generative models, we need to check for cache only when present

File: src/transformers/models/evolla/modeling_evolla.py (hunk 11 above)
Comment (zucchini-nlp) [reply src/transformers/models/evolla/modeling_evolla.py:243]: This "eager_attention"  is used twice in two different attention classes in evolla, and the other attention has no position embedding types

Didn't want to have two different `eager_attn` for each module

Comment (vasqu) [reply src/transformers/models/evolla/modeling_evolla.py:243]: Ah ic, then we can take the above slightly modified? Makes sense then

File: src/transformers/utils/generic.py [hunk 14]
```diff
@@ -974,22 +974,22 @@ def check_model_inputs(func):
...
+            kwargs["use_cache"] if kwargs.get("use_cache") is not None else getattr(self.config, "use_cache", None)
+        )
+        if use_cache is not None:
+            if getattr(self, "gradient_checkpointing", False) and self.training and use_cache:
+                logger.warning_once(
+                    "`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`."
+                )
+                use_cache = False
+
+            kwargs["use_cache"] = use_cache
```
Comment (vasqu) [inline src/transformers/utils/generic.py:987]: let's keep that line one level below (not under the if), my guts tell me we might break something but not having a rationale here

File: src/transformers/models/evolla/modular_evolla.py [hunk 15]
```diff
@@ -193,6 +194,17 @@ class EvollaSaProtPooler(EsmPooler):
 class EvollaSaProtPreTrainedModel(PreTrainedModel):
     config: SaProtConfig
     _no_split_modules = ["EvollaSaProtLayer"]
+    _supports_flash_attn = True
+    _supports_sdpa = True
+    _supports_attention_backend = True
```
Comment (vasqu) [inline src/transformers/models/evolla/modular_evolla.py:200]: we can add `_supports_flex_smthn` here too I think

File: src/transformers/models/evolla/modeling_evolla.py [hunk 16]
```diff
@@ -229,11 +221,61 @@ def forward(self, q: torch.Tensor, k: torch.Tensor) -> tuple[torch.Tensor, torch
...
+    **kwargs: Unpack[TransformersKwargs],
+):
+    # EVOLLA_SA_PROT applies relative position embeddings and we don't copy from Llama
+    attn_weights = torch.matmul(query, key.transpose(2, 3)) * scaling
+
+    if attention_mask is not None:
+        causal_mask = attention_mask[:, :, :, : key.shape[-2]]
+        attn_weights = attn_weights + causal_mask
+
+    if hasattr(module, "position_embedding_type") and module.position_embedding_type in [
```
Comment (vasqu) [inline src/transformers/models/evolla/modeling_evolla.py:243]: Just to be safe let's keep the order:
- matmul
- position scores
- scaling
- mask
- ....

Just don't want to break things even if this is prolly low usage

**Review by vasqu — APPROVED** [review]
Just one small nit (order of relative scaling) and could you change the mask creation per https://github.com/huggingface/transformers/pull/40370/files#r2297912836

LGTM otherwise! cc @pstjohn re #40211

File: src/transformers/utils/generic.py (hunk 14 above)
Comment (zucchini-nlp) [reply src/transformers/utils/generic.py:987]: you mean dedent `use_cache` one tab back? It will cause issues with non-generative models and tbh we should've checked for this condition in prev PRs

File: src/transformers/models/evolla/modular_evolla.py (hunk 15 above)
Comment (zucchini-nlp) [reply src/transformers/models/evolla/modular_evolla.py:200]: yeah, might be if doesn't throw issues, will check

File: src/transformers/models/evolla/modeling_evolla.py (hunk 16 above)
Comment (zucchini-nlp) [reply src/transformers/models/evolla/modeling_evolla.py:243]: it has no scaling applied in main branch version so it is a no-op with 1 😄

File: src/transformers/utils/generic.py (hunk 14 above)
Comment (vasqu) [reply src/transformers/utils/generic.py:987]: Nope, the `kwargs["use_cache"] = use_cache` a tab back. Pretty sure it wasn't under the if before, e.g. https://github.com/huggingface/transformers/blob/5a8ba87ecf17aecec428a67e96cb14f8a9a74055/src/transformers/utils/generic.py#L985-L991 (last line)

Comment (zucchini-nlp) [reply src/transformers/utils/generic.py:987]: moving it back means `use_cache` will be passed as an argument to all models' where `can_record_output` is set. We can't do it, some models will throw errors that an unexpected arg is passed

The check I added is robust enough as we pass `use_cache` only when the config has it, i.e. in all generative models

Comment (vasqu) [reply src/transformers/utils/generic.py:987]: Omg, yea my bad 🤦‍♂️

**Comment by zucchini-nlp** [comment]
run-slow: esm, evolla

**Comment by zucchini-nlp** [comment]
run-slow: esm, evolla

File: src/transformers/models/esm/modeling_esm.py [hunk 17]
```diff
@@ -340,6 +340,7 @@ def __init__(self, config, position_embedding_type=None, layer_idx=None):
         self.scaling = 1.0  # For BC we apply scaling before RoPE
         self.is_decoder = config.is_decoder
         self.layer_idx = layer_idx
+        self.is_causal = self.is_decoder  # used only in FA2/FA3
```
Comment (vasqu) [inline src/transformers/models/esm/modeling_esm.py]: Sdpa uses this as well :D this is probably incorrect in case of cross-attention, can you add an optional argument here instead

Comment (zucchini-nlp) [reply src/transformers/models/esm/modeling_esm.py]: Ah right, it also has cross attention. Copied from current ESM attention assuming it would be working on main branch
